src/transformStateWithClones.js

```diff
@@ -8,6 +8,33 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const stateHistory = [];
+  let previousState = { ...state };
+
+  for (const action of actions) {
+    const { type } = action;
+
+    let newState = { ...previousState };
+
+    if (type === 'addProperties') {
+      newState = Object.assign({}, previousState, action.extraData);
+    }
+
+    if (type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    }
+
+    if (type === 'clear') {
+      newState = {};
+    }
```

Comment:
[CODE STYLE]: A switch/case statement should always have a default case for error handling. This helps to catch unexpected values of `type` and handle them appropriately.